### PR TITLE
Support literals for field names.

### PR DESCRIPTION
**Pull request by rbtcollins**
This permits : in field names but also potentially anything
stringifiable, which may be overly liberal.

Signed-off-by: Robert Collins <robert.collins@cognite.com>

## Motivation

Some opentracing systems use fields with : in the names, which are not legal rust identifiers. We could special case :, but then we have complicated double-nested patterns (repeated idents separated by : and a nested repeat separated by .), and this may by a case of always being one-step-behind as more cases turn up; so this patch instead just gets in front and lets users put in whatever they want: as they are not rust identifiers, they miss out on some niceness.

### CODE REVIEW

File: tracing/src/macros.rs [hunk 1]
```diff
@@ -1959,6 +1959,52 @@ macro_rules! valueset {
             $next,
         )
     };
+
+    // TODO(#1138): determine a new syntax for uninitialized span fields, and
+    // re-enable this.
+    // (@{ $(,)* $($out:expr),* }, $next:expr, $k:literal = _, $($rest:tt)*) => {
+    //     $crate::valueset!(@ { $($out),*, (&$next, None) }, $next, $($rest)*)
+    // };
```
Comment (hawkw) [inline tracing/src/macros.rs:1967]: This is no longer necessary --- disabled fields can be created with `field::Empty`. We should probably not copy-paste this comment any more (and I should, eventually, remove the original).

File: tracing/src/macros.rs [hunk 2]
```diff
@@ -2017,6 +2063,21 @@ macro_rules! fieldset {
...
+        $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($k) } $($rest)*)
+    };
+    (@ { $(,)* $($out:expr),* } $k:literal = $val:expr, $($rest:tt)*) => {
+        $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($k) } $($rest)*)
+    };
+    // TODO(#1138): determine a new syntax for uninitialized span fields, and
+    // re-enable this.
+    // (@ { $($out:expr),* } "$k:ident" = _, $($rest:tt)*) => {
+    //     $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($($k).+) } $($rest)*)
+    // };
```
Comment (hawkw) [inline tracing/src/macros.rs:2079]: Similarly, we don't need to keep repeating this.

File: README.md [hunk 3]
```diff
@@ -154,14 +154,17 @@ pub fn shave(yak: usize) -> Result<(), Box<dyn Error + 'static>> {
...
     //
     // let span = span!(Level::TRACE, "shaving_yaks", yaks = yaks);
     //
     // local variables (`yaks`) can be used as field values
     // without an assignment, similar to struct initializers.
-    let span = span!(Level::TRACE, "shaving_yaks", yaks);
+    //
+    // Also note the quotation marks for the guid:x-request-id, needed because
+    // it is not a simple token.
+    let span = span!(Level::TRACE, "shaving_yaks", "guid:x-request-id" = "abcdef", yaks);
```
Comment (hawkw) [inline README.md:167]: I don't think it's necessary to show the syntax for literal field names in the basic usage example in the README. This example is kind of intended to demonstrate the basics of using `tracing`, and I'd prefer to show more advanced usage elsewhere.

It's probably better to show this usage in the documentation for the `span!` and `event!` macros, which you've already added,, and maybe add an example in the `examples` dir as well.

File: tracing/src/lib.rs [hunk 4]
```diff
@@ -287,6 +287,17 @@
...
 //!
+//! Fields may have :'s in their names, but this may not be used with the local
+//! variable shorthand as : is not a valid rust pathname.
+//! ```
+//! # use tracing::{span, Level};
+//! # fn main() {
+//! // records an event with an : containing fieldname:
+//! //  - "guid:x-request-id", with the value abcdef
+//! span!(Level::TRACE, "api", "guid:x-request-id" = "abcdef");
+//! # }
```
Comment (hawkw) [inline tracing/src/lib.rs:298]: Rather than discussing colons in particular, I might reword this to just discuss fields that are not valid Rust identifiers, in general:
```suggestion
//! Fields with names that are not Rust identifiers, or with names that are Rust reserved words,
//! may be created using quoted string literals. However, this may not be used with the local
//! variable shorthand.
//! ```
//! # use tracing::{span, Level};
//! # fn main() {
//! // records an event with fields whose names are not Rust identifiers
//! //  - "guid:x-request-id", containing a `:`, with the value "abcdef"
//! //  - "type", which is a reserved word, with the value "request"
//! span!(Level::TRACE, "api", "guid:x-request-id" = "abcdef", "type" = "request");
//! # }
```

File: tracing/src/macros.rs [hunk 5]
```diff
@@ -1959,6 +1959,52 @@ macro_rules! valueset {
             $next,
         )
     };
+
```
Comment (hawkw) [inline tracing/src/macros.rs:1962]: Might be worth a comment here to separate the cases with literal field names from the cases with ident field names.

File: tracing/src/macros.rs [hunk 6]
```diff
@@ -2017,6 +2063,21 @@ macro_rules! fieldset {
         $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($($k).+) } $($rest)*)
     };
 
+    (@ { $(,)* $($out:expr),* } $k:literal = ?$val:expr, $($rest:tt)*) => {
```
Comment (hawkw) [inline tracing/src/macros.rs:2066]: Again, maybe worth adding a comment to separate the two blocks of cases.

File: tracing/src/macros.rs [hunk 7]
```diff
@@ -2017,6 +2063,21 @@ macro_rules! fieldset {
...
+        $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($k) } $($rest)*)
+    };
+    (@ { $(,)* $($out:expr),* } $k:literal = %$val:expr, $($rest:tt)*) => {
+        $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($k) } $($rest)*)
+    };
+    (@ { $(,)* $($out:expr),* } $k:literal = $val:expr, $($rest:tt)*) => {
+        $crate::fieldset!(@ { $($out),*, $crate::__tracing_stringify!($k) } $($rest)*)
+    };
```
Comment (hawkw) [inline tracing/src/macros.rs:2074]: I don't think this is correct. The `stringify!` compiler built-in macro is used here to convert tokens that are identifiers (or dotted sequences of identifiers) to string literals. Using `stringify!` on a token that's already a string literal will create a string literal containing the _string representation_ of that token: it will contain quotation marks and escape sequences for any escaped characters. For example, see [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4760fe1b9865355171deb81aac740bbc).

I think instead of stringifying fields whose names are already string literals, we just want to use the provided string literal directly:
```suggestion
    (@ { $(,)* $($out:expr),* } $k:literal = ?$val:expr, $($rest:tt)*) => {
        $crate::fieldset!(@ { $($out),*, $k } $($rest)*)
    };
    (@ { $(,)* $($out:expr),* } $k:literal = %$val:expr, $($rest:tt)*) => {
        $crate::fieldset!(@ { $($out),*, $k } $($rest)*)
    };
    (@ { $(,)* $($out:expr),* } $k:literal = $val:expr, $($rest:tt)*) => {
        $crate::fieldset!(@ { $($out),*, $k) } $($rest)*)
    };
```

File: tracing/src/macros.rs (hunk 1 above)
Comment (rbtcollins) [reply tracing/src/macros.rs:1967]: Removed

File: tracing/src/macros.rs (hunk 2 above)
Comment (rbtcollins) [reply tracing/src/macros.rs:2079]: Removed.

File: README.md (hunk 3 above)
Comment (rbtcollins) [reply README.md:167]: I've added an example, its hopefully ok.

File: tracing/src/lib.rs (hunk 4 above)
Comment (rbtcollins) [reply tracing/src/lib.rs:298]: done

File: tracing/src/macros.rs (hunk 5 above)
Comment (rbtcollins) [reply tracing/src/macros.rs:1962]: done

File: tracing/src/macros.rs (hunk 6 above)
Comment (rbtcollins) [reply tracing/src/macros.rs:2066]: done

File: tracing/src/macros.rs (hunk 7 above)
Comment (rbtcollins) [reply tracing/src/macros.rs:2074]: This suggestion doesn't compile

Comment (rbtcollins) [reply tracing/src/macros.rs:2074]: There's a typo in the suggestion, but yes, done.

File: examples/examples/attrs-literal-field-names.rs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+#![deny(rust_2018_idioms)]
+
+use tracing::{debug, info, span, Level};
```
Comment (hawkw) [inline examples/examples/attrs-literal-field-names.rs:3]: Let's remove the `info` import, as we don't use it in the example:
```suggestion
use tracing::{debug, span, Level};
```

File: examples/examples/attrs-literal-field-names.rs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+#[instrument]
+#[inline]
+fn suggest_band() -> String {
+    debug!("Suggesting a band.");
+    String::from("Wild Pink")
+}
+
+fn main() {
+    let subscriber = tracing_subscriber::fmt()
+        .with_env_filter("attrs_basic=trace")
```
Comment (hawkw) [inline examples/examples/attrs-literal-field-names.rs:15]: This filter won't enable anything, since the name of this module is `attrs-literal-field-names`:
```suggestion
        .with_env_filter("attrs_literal_field_names=trace")
```

File: examples/examples/attrs-literal-field-names.rs (hunk 8 above)
Comment (rbtcollins) [reply examples/examples/attrs-literal-field-names.rs:3]: done

File: examples/examples/attrs-literal-field-names.rs (hunk 9 above)
Comment (rbtcollins) [reply examples/examples/attrs-literal-field-names.rs:15]: done

File: tracing/tests/macros.rs [hunk 10]
```diff
@@ -262,6 +262,22 @@ fn error_span_with_parent() {
     error_span!(parent: &p, "bar",);
 }
 
+#[test]
+fn span_with_non_rust_symbol() {
```
Comment (hawkw) [inline tracing/tests/macros.rs:266]: May be good to add a similar compile test for an event, just in case. This is _probably_ sufficient, since a regression would probably be in the shared key-value parsing macros, so it may not matter a whole lot...

Comment (rbtcollins) [reply tracing/tests/macros.rs:266]: We only change valueset and fieldset, not span or event directly, so I think just testing span! is fine, unless things get rearranged radically.

**Comment by rbtcollins** [comment]
> Looks good to me overall. Mind adding a description of the new example to the list of examples?

done

**Comment by hawkw** [comment]
And merged! Thanks for working on this!

**Comment by rbtcollins** [comment]
You're welcome. Thank you for merging it. :)
